File: src/Cyborg/Systems/PlayerSystem.cs
using System;
using System.Linq;
using Cyborg.Components;
using Cyborg.Core;
using Cyborg.Entities;
using Cyborg.Utilities;
using Microsoft.Xna.Framework;

namespace Cyborg.Systems
{
    public class PlayerSystem : IUpdateSystem
    {
        private const float _walkingForce = 600f;
        private const float _attackDuration = 0.25f;
        private const float _attackKnockbackVelocity = 400f;
        private const float _dashDuration = 0.75f;
        private const float _dashInstantaneousVelocity = 400f;
        private readonly IEntityManager _entityManager;
        private readonly IGameState _gameState;

        public PlayerSystem(IEntityManager entityManager, IGameState gameState)
        {
            _entityManager = entityManager;
            _gameState = gameState;
        }

        public void Update(GameTime gameTime)
        {
            if (!_gameState.Active)
                return;

            var elapsed = (float)gameTime.ElapsedGameTime.TotalSeconds;

            foreach (var entity in _entityManager.Entities<IPlayer>())
            {
                CheckStateChange(entity);
                ApplyState(entity, elapsed);
                ApplyAnimation(entity);
            }
        }

        private static void CheckStateChange(IPlayer entity)
        {
            // Dash
            if (entity.Player.Dashing && entity.Player.DashElapsed >= _dashDuration)
                entity.Player.Dashing = false;

            if (!entity.Player.Dashing && !entity.Player.Attacking && entity.Controller.Pressed.Contains(Button.Dash))
            {
                entity.Player.Dashing = true;
                entity.Player.DashElapsed = 0f;

                if (entity.Controller.Joystick != Vector2.Zero)
                    entity.Player.Direction = entity.Controller.Joystick;

                entity.Kinetic.Velocity = entity.Player.Direction * _dashInstantaneousVelocity;
            }

            // Attack
            if (entity.Player.Attacking && entity.Player.AttackElapsed >= _attackDuration)
                entity.Player.Attacking = false;

            if (!entity.Player.Attacking && entity.Controller.Pressed.Contains(Button.Attack))
            {
                entity.Player.Attacking = true;
                entity.Player.AttackElapsed = 0f;
                entity.Player.AttackCounter++;

                if (entity.Controller.Joystick != Vector2.Zero)
                    entity.Player.Direction = entity.Controller.Joystick;

                var cardinalDirection = entity.Player.Direction.ToCardinal();
                double baseAttackAngle = 0;
                if (cardinalDirection.X == 1)
                    baseAttackAngle = 0;
                else if (cardinalDirection.X == -1)
                    baseAttackAngle = Math.PI;
                else if (cardinalDirection.Y == 1)
                    baseAttackAngle = 0.5 * Math.PI;
                else if (cardinalDirection.Y == -1)
                    baseAttackAngle = 1.5 * Math.PI;

                var adjustedAttackAngle = baseAttackAngle + 0.125 * Math.PI * ((entity.Player.AttackCounter % 2 == 0) ? 1 : -1);
                var attackSector = (adjustedAttackAngle - 0.3125 * Math.PI, adjustedAttackAngle + 0.3125 * Math.PI);
                entity.Player.AttackAngles = attackSector;
            }

            // Walk
            if (!entity.Player.Walking && !entity.Player.Attacking && !entity.Player.Dashing && entity.Controller.Joystick != Vector2.Zero)
                entity.Player.Walking = true;

            if (entity.Player.Walking && (entity.Player.Attacking || entity.Player.Dashing || entity.Controller.Joystick == Vector2.Zero))
                entity.Player.Walking = false;
        }

        private void ApplyState(IPlayer entity, float elapsed)
        {
            // Dash
            if (entity.Player.Dashing)
                entity.Player.DashElapsed += elapsed;

            // Attack
            if (entity.Player.Attacking)
            {
                entity.Player.AttackElapsed += elapsed;

                var attackBounds = new Sector(entity.Body.Position.ToRoundedPoint(), entity.Player.AttackRadius, entity.Player.AttackAngles.Minimum, entity.Player.AttackAngles.Maximum);
                foreach (var enemyEntity in _entityManager.Entities<IEnemy>())
                {
                    if (!attackBounds.Intersects(enemyEntity.Body.Bounds))
                        continue;

                    if (!enemyEntity.Damage.TryApply(1))
                        continue;

                    var knockbackVector = Vector2.Normalize(enemyEntity.Body.Position - entity.Body.Position);
                    enemyEntity.Kinetic.Velocity = knockbackVector * _attackKnockbackVelocity;
                }
            }

            // Walk
            if (entity.Player.Walking)
            {
                if (entity.Controller.Joystick != Vector2.Zero)
                    entity.Player.Direction = entity.Controller.Joystick;

                entity.Kinetic.Force = entity.Player.Direction * _walkingForce;
            }
            else
                entity.Kinetic.Force = Vector2.Zero;
        }

        private static void ApplyAnimation(IPlayer entity)
        {
            var cardinalDirection = entity.Player.Direction.ToCardinal();
            if (entity.Player.Attacking)
            {
                if (cardinalDirection.X == 1)
                    entity.Sprite.Animation = (entity.Player.AttackCounter % 2 == 0) ? CyborgHero.AnimationAttackRight : CyborgHero.AnimationAttack2Right;
                else if (cardinalDirection.X == -1)
                    entity.Sprite.Animation = (entity.Player.AttackCounter % 2 == 0) ? CyborgHero.AnimationAttackLeft : CyborgHero.AnimationAttack2Left;
                else if (cardinalDirection.Y == 1)
                    entity.Sprite.Animation = (entity.Player.AttackCounter % 2 == 0) ? CyborgHero.AnimationAttackDown : CyborgHero.AnimationAttack2Down;
                else if (cardinalDirection.Y == -1)
                    entity.Sprite.Animation = (entity.Player.AttackCounter % 2 == 0) ? CyborgHero.AnimationAttackUp : CyborgHero.AnimationAttack2Up;
            }
            else if (entity.Player.Walking)
            {
                if (cardinalDirection.X == 1)
                    entity.Sprite.Animation = CyborgHero.AnimationWalkRight;
                else if (cardinalDirection.X == -1)
                    entity.Sprite.Animation = CyborgHero.AnimationWalkLeft;
                else if (cardinalDirection.Y == 1)
                    entity.Sprite.Animation = CyborgHero.AnimationWalkDown;
                else if (cardinalDirection.Y == -1)
                    entity.Sprite.Animation = CyborgHero.AnimationWalkUp;
            }
            else
            {
                if (cardinalDirection.X == 1)
                    entity.Sprite.Animation = CyborgHero.AnimationStandRight;
                else if (cardinalDirection.X == -1)
                    entity.Sprite.Animation = CyborgHero.AnimationStandLeft;
                else if (cardinalDirection.Y == 1)
                    entity.Sprite.Animation = CyborgHero.AnimationStandDown;
                else if (cardinalDirection.Y == -1)
                    entity.Sprite.Animation = CyborgHero.AnimationStandUp;
            }
        }
    }
}

File: src/systems/playerSystem.ts
import { System } from ".";
import { AnimatedSpriteSet, hasBody, hasPhysics, hasSprite, isPlayer } from "../components";
import camera from "../framework/camera";
import gameState from "../framework/gameState";
import { getInput } from "../framework/input";
import { Vector, cardinalise, hasValue, multiply, normalise } from "../shapes";

const walkingForce = 600;
const attackDuration = 0.25;
const attackKnockbackVelocity = 400;
const dashDuration = 0.75;
const dashInstantaneousVelocity = 400;

const playerSystem: System = (entities, _, deltaSeconds) => {
	if (!gameState.active()) return;

	for (const entity of entities) {
		if (!isPlayer(entity)) continue;

		const playerPosition = hasBody(entity) ? entity.position : { x: camera.width / 2, y: camera.height / 2 };
		const input = getInput(playerPosition);

		// Check state end
		if (entity.dashing.active && entity.dashing.elapsed >= dashDuration) entity.dashing.active = false;
		if (entity.attacking.active && entity.attacking.elapsed >= attackDuration) entity.attacking.active = false;
		if (entity.walking.active && (entity.attacking.active || entity.dashing.active || !hasValue(input.moveDirection))) entity.walking.active = false;

		// Check state start
		if (!entity.dashing.active && !entity.attacking.active && input.dash) {
			entity.dashing.active = true;
			entity.dashing.elapsed = 0;
			if (hasValue(input.moveDirection)) entity.direction = input.moveDirection;
			if (hasPhysics(entity)) entity.velocity = multiply(entity.direction, dashInstantaneousVelocity);
		}
		if (!entity.attacking.active && input.attack) {
			entity.attacking.active = true;
			entity.attacking.elapsed = 0;
			entity.attacking.counter++;
			if (hasValue(input.moveDirection)) entity.direction = input.moveDirection;
		}
		if (!entity.walking.active && !entity.dashing.active && !entity.attacking.active && hasValue(input.moveDirection)) entity.walking.active = true;

		// Apply state
		if (entity.dashing.active) entity.dashing.elapsed += deltaSeconds;
		if (entity.attacking.active) entity.attacking.elapsed += deltaSeconds;
		if (entity.walking.active) {
			entity.direction = input.moveDirection;
			if (hasPhysics(entity)) entity.acceleration = multiply(entity.direction, walkingForce);
		} else {
			if (hasPhysics(entity)) entity.acceleration = { x: 0, y: 0 };
		}

		// Apply animation
		if (!hasSprite(entity) || !(entity.sprite instanceof AnimatedSpriteSet)) continue;

		const directionSuffix = toDirectionString(entity.direction);
		if (entity.attacking.active) {
			const attackSuffix = entity.attacking.counter % 2 ? "alt" : "";
			entity.sprite.play(`attack${directionSuffix}${attackSuffix}`);
		} else if (entity.walking.active) entity.sprite.play(`walk${directionSuffix}`);
		else entity.sprite.play(`stand${directionSuffix}`);
	}
};

const toDirectionString = (direction: Vector): string => {
	const cardinal = hasValue(direction) ? cardinalise(direction) : { x: 0, y: 0 };
	if (cardinal.x === 1 && cardinal.y === 0) return "right";
	else if (cardinal.x === -1 && cardinal.y === 0) return "left";
	else if (cardinal.x === 0 && cardinal.y === 1) return "down";
	else if (cardinal.x === 0 && cardinal.y === -1) return "up";
	else return "down";
};

export default playerSystem;
